onnxruntime/test/providers/cpu/tensor/grid_sample_test.cc

```diff
@@ -25,8 +34,7 @@ TEST(GridsampleTest, test_grid_sample_16_4D_nearest_zeros_align_corners) {
   test.AddAttribute("padding_mode", padding_mode);
   test.AddAttribute("align_corners", align_corners);
   test.AddOutput<float>("Y", Y_shape, Y_data);
-  test.ConfigEp(DefaultCpuExecutionProvider())
-      .RunWithConfig();
+  test.Run(OpTester::ExpectResult::kExpectSuccess, "", GetExcludedExecutionProviders(16));
```

Comment:
Suggest to use ConfigEps like the following:
```
  std::vector<std::unique_ptr<IExecutionProvider>> execution_providers;
  execution_providers.emplace_back(DefaultCpuExecutionProvider());
#ifdef USE_CUDA
  execution_providers.emplace_back(DefaultCudaExecutionProvider());
#endif
  test.ConfigEps(std::move(execution_providers)).RunWithConfig();
 ```
 
Need change the test to cover both NCHW and NHWC: 
if ENABLE_CUDA_NHWC_OPS is defined, test NHWC like the following: transpose X, and Y, change operator domain to internal NHWC domain, then run the test again.

Reply:
This change already improves GridSample testing a lot on all EPs and even found one bug in the current CUDA implementation. 

To get full NHWC testing without having to modify all tests the best idea I have come up so far is add a new Cuda NHWC EP which sets prefer_nwhc = true in the factory. This way all tests can run with the CUDA EP and the CUDA NHWC EP automatically. I haven't verified yet if this will break any CUDA EP based special handling.

Reply:
With regards to your proposal, before I tested the same idea by passing the `execution_providers` vector to `BaseTester::Run()` with the result that only the first EP in the list is executed.

The same thing happens with your proposal passing a vector of execution providers to the `ConfigEps` function. If this is supposed to work I'll file an issue.

To get this PR integrated my proposal is that I'll add all known EPs except for the CUDA EP and CPU EP into the excluded list.

Unfortunately it is not possible to have two CUDA EPs with two different configurations in the list for testing. Thus it's not possible to enable NHWC testing by default for now. I'll take a closer look if there is an easy workaround for this issue.

Reply:
You are right that only one provider is used in testing (the first one that has the implementation) so we add CUDA EP to the list first before CPU provider. In this way, CUDA provider is tested in cuda build, and CPU provider is tested in CPU build.

Another way is to test each provider separately like [this](https://github.com/microsoft/onnxruntime/blob/1b48054e1b7991ccef664fbedd659ec95d0e7ca7/onnxruntime/test/contrib_ops/attention_op_test.cc#L230-L252)

Reply:
I remember that I tried to test something like this as well and had problems with the failure report. I'll give it another try. Eventually it'd be good if the tester will handle all implementations for us to avoid code duplication.

The reason why I wanted to add the CUDA EP to this test is avoid code duplication. The NHWC tests are extremely incomplete and it doesn't make sense to copy/paste this test to the NHWC folder and the cuda test folder as well.

In my ideal world all EPs will get the same amount of testing to have the most consistent result.

Rewriting all test base class it out of scope for this CL.

My suggestion is that I'll remove the changed test and we run the same 'bad' tests as today for the GridSample implementation. In another change I'll address the issues of running on test with multiple EPs in a positive list and also allow running the same EP with different options to have the chance for even more complete testing.

Reply:
It's fine to use current change of this file to test NCHW cuda implementation.

My major concern is test coverage. Currently, this test does not cover NHWC. If you need test NHWC, need add domain like `OpTester test("GridSample", 16, kMSInternalNHWCDomain)`

The NHWC tests compares NCHW and NHWC results. I think we need at least one such test to ensure NHWC code is tested.

Reply:
I have implemented a small test function which run a test over a list of specified test. If probably could be added to the testing infrastructure in the future.

Test will run for NHWC & NCHW for the CUDA EP.